git-server/src/lib.rs

```diff
@@ -24,19 +29,74 @@ pub struct Options {
     pub tls_cert: Option<PathBuf>,
     pub tls_key: Option<PathBuf>,
     pub git_receive_pack: bool,
+    pub authorized_keys: Vec<String>,
+    pub cert_nonce_seed: Option<String>,
+    pub allow_unauthorized_keys: bool,
+    pub hook_logs_dir: Option<String>,
+    pub log_hooks: bool,
 }
 
 #[derive(Debug, Clone)]
 pub struct Context {
     root: PathBuf,
     git_receive_pack: bool,
+    authorized_keys: Vec<String>,
+    cert_nonce_seed: Option<String>,
+    allow_unauthorized_keys: bool,
+    hook_logs_dir: Option<String>,
+    log_hooks: bool,
+}
+
+impl Context {
+    // Sets the config receive.advertisePushOptions, which lets the user known they can provide a push option `-o`,
+    // to specify unique attributes. This is currently not used, but may be used in the future.
+    pub fn advertise_push_options(&self) -> Result<(), Error> {
+        let field = "receive.advertisePushOptions";
+        let value = "true";
+
+        self.set_root_git_config(field, value)?;
+
+        Ok(())
+    }
+
+    // Enables users to submit a signed push: `push --signed`
+    pub fn set_cert_nonce_seed(&self) -> Result<(), Error> {
+        let field = "receive.certNonceSeed";
+        let value = self
+            .cert_nonce_seed
+            .clone()
+            .unwrap_or_else(gen_random_string);
+
+        self.set_root_git_config(field, &value)?;
+
+        Ok(())
+    }
+
+    // updates the git config in the root project
+    pub fn set_root_git_config(&self, field: &str, value: &str) -> Result<(), Error> {
+        let mut path = self.root.clone();
+        path.push("git/config");
+
+        tracing::debug!("searching for git config at: {:?}", path);
+
+        let mut config = git2::Config::open(&path)?;
+
+        config.set_str(field, value)?;
+
+        Ok(())
+    }
```

Comment:
The functions above should use `///` docstrings

Reply:
5030fa7 updates